src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -146,43 +145,38 @@ impl PushDownProjection {
                 let [required_columns] = &plan.required_columns()[..] else {
                     panic!()
                 };
-                if required_columns.len() < upstream_schema.names().len() {
-                    let pruned_upstream_schema = upstream_schema
-                        .fields
-                        .iter()
-                        .filter_map(|(name, field)| {
-                            required_columns.contains(name).then(|| field.clone())
-                        })
-                        .collect::<Vec<_>>();
-                    let schema = Schema::new(pruned_upstream_schema)?;
-                    let new_source: LogicalPlan = match source.source_info.as_ref() {
-                        SourceInfo::ExternalInfo(ExternalInfo::Scan(scan_external_info)) => {
-                            Source::new(
+                match source.source_info.as_ref() {
+                    SourceInfo::ExternalInfo(external_info) => {
+                        if required_columns.len() < upstream_schema.names().len() {
+                            let pruned_upstream_schema = upstream_schema
+                                .fields
+                                .iter()
+                                .filter_map(|(name, field)| {
+                                    required_columns.contains(name).then(|| field.clone())
+                                })
+                                .collect::<Vec<_>>();
+                            let schema = Schema::new(pruned_upstream_schema)?;
+                            let new_source: LogicalPlan = Source::new(
                                 schema.into(),
-                                Arc::new(SourceInfo::ExternalInfo(ExternalInfo::Scan(
-                                    ScanExternalInfo {
-                                        pushdowns: scan_external_info.pushdowns.with_columns(Some(
-                                            Arc::new(required_columns.iter().cloned().collect()),
-                                        )),
-                                        ..scan_external_info.clone()
-                                    },
+                                Arc::new(SourceInfo::ExternalInfo(external_info.with_pushdowns(
+                                    external_info.pushdowns().with_columns(Some(Arc::new(
+                                        required_columns.iter().cloned().collect(),
+                                    ))),
                                 ))),
-                                source.limit,
                             )
-                            .into()
+                            .into();
+                            let new_plan = plan.with_new_children(&[new_source.into()]);
+                            // Retry optimization now that the upstream node is different.
```

Comment:
Why is this necessary/what are the failure modes here?

Reply:
This recursive call is used to [prune now-redundant projections](https://github.com/Eventual-Inc/Daft/blob/0c450a535065a7fe42f744c372898dcf45e8883b/src/daft-plan/src/optimization/rules/push_down_projection.rs#L33-L54). This (status quo) behavior currently assumes that if a projection is pushed into a scan, then the downstream projection op can be removed from the logical plan.

I thought about this a bit, and I do think that we'll want to eventually integrate this with a query to the `ScanOperator` to see if fully absorbing the projection is supported, but I don't think that this is as needed as it is for limit and filter pushdowns, right? I think that all of our scan operators will support fully absorbing projections.

Reply:
> support  fully absorbing projections

ScanOperators will only be able to fully support projects that don't have any transformations I guess (pure column pruning). If the projection contains any transformation at all, we can't absorb...

Reply:
Yep, our projection pushdown is already very conservative, where we only drop a projection if it's a raw column selection projection! Expanding these semantics when we add more nontrivial projection pushdowns into scans seems doable.